include/bitcoin/database/impl/hash_table.ipp

```diff
@@ -93,14 +93,17 @@ template <typename Manager, typename Index, typename Link, typename Key>
 typename hash_table<Manager, Index, Link, Key>::const_value_type
 hash_table<Manager, Index, Link, Key>::find(Link link) const
 {
+    if (manager_.past_eof(link))
+        return { manager_, not_found, list_mutex_ };
+
     return { manager_, link, list_mutex_ };
```

Comment:
We should consistently make the same assumption about links (i.e. that they are always valid). As such this should be:
```cpp
    // Ensure requested position is within the file.
    // We avoid a runtime error here to optimize out the past_eof locks.
    BITCOIN_ASSERT_MSG(!manager_.past_eof(link), "Read past end of file.");

    return { manager_, link, list_mutex_ };
```

Reply:
I agree with the change. I wanted to put in an assertion as well instead of the the return.

But usage like the following confused me a bit:
```
bool transaction_database::candidize(link_type link, bool positive)
{
   const auto element = hash_table_.find(link);

    if (!element)
        return false;
```

There are a couple more places that returned element from find is essentially checked for not_found sentinel. But with the current implementation of find, the `!element` test is redundant, as it always returns an element with link passed in.

I wonder if we make the change you suggested (and I prefer that change) then should we change the usage like above to the following?

```
bool transaction_database::candidize(link_type link, bool positive)
{
    if (link == not_found)
        return false;

   const auto element = hash_table_.find(link);
```

Or can we completely take out the `!element` test?

Maybe I am missing something here?

Reply:
It's possible that the above test for a not_found link allows for passing in a not_found link sentinel.

Reply:
> I think the issue is that find(hash) returns a possible not_found element and the find(link) method is testing for the result as a matter of consistency.

You hit the nail on the head there.

I did a quick review of calls to find(link) making sure:

1. We don't call find(link) with a not_found sentinel and instead terminate the code path earlier.

2. Given that find(link) will no longer returns an not_found element, I removed the checks for not_found element that created the confusion for me in the first place.

The changes are in the latest commit to this PR (https://github.com/libbitcoin/libbitcoin-database/pull/257/commits/a66f91aa2f893d4cd8418b10abc724d18243aa04)